### ISSUE TITLE
No way to describe ConfigDescriptor without path

### ISSUE THREAD

**Issue by senia-psm** [zio/zio-config#258]
There is no way to describe ConfigDescriptor without path.
So in case of complicated map elements there is code duplication:
In the simplest case it would be something like this:
```scala
list(string("path").orElseEither(int("path")))
```
For more complicated case, like "treat list of strings as list of case objects in sealed class hierarchy" there would be a lot more cases with `orElse`.

This issue related to #257 - it's not so easy to create documentation for such descriptors.

Use case:
```scala
sealed trait X
case object A           extends X
case object B           extends X
case object C           extends X
case class D(s: String) extends X

case class ConstantString(value: String) extends PropertyType[String, String] {
  def read(propertyValue: String): Either[PropertyType.PropertyReadError[String], String] =
    if (propertyValue == value) Right(value)
    else Left(PropertyType.PropertyReadError(propertyValue, s"constant string '$value'"))
  def write(a: String): String = a
}

def constant[T](label: String, value: T)(path: String): ConfigDescriptor[String, String, T] =
  ConfigDescriptor.Source(path, ConfigSource.empty, ConstantString(label))(
    _ => value,
    (p: T) => Some(p).filter(_ == value).map(_ => label)
  ) ?? s"constant string '$label'"

list(
  constant[X]("a", A)("path")
    .orElse(constant[X]("b", B)("path"))
    .orElse(constant[X]("c", C)("path"))
    .orElse(nested("path")(nested("d")(string("s")(D, D.unapply))))
)
```

**Comment by afsalthaj** [comment]
Map isn’t implemented yet. You wouldn’t need to implement by yourself on this.

For very complicated case, use magnolia. 

@jdegoes would you like to add some points to this

**Comment by senia-psm** [comment]
@afsalthaj We just discussed it on discord and there is no way to implement it with current `DeriveConfigDescriptor`

**Comment by afsalthaj** [comment]
@senia-psm 

Let's describe config for list of element without path to list, not using zio-config, but just using the basic typesafe-config library.
https://github.com/lightbend/config. 

```scala

  val s =
    s"""
       |[
       |  {
       |   x = 1
       |  }
       |
       |]
       |""".stripMargin

ConfigFactory.parseString(s)

```


By default (meaning, using the apis), 
typesafe-config gives a failure for the above config.

The following is the error

```scala
Exception in thread "main" com.typesafe.config.ConfigException$WrongType: String: 2:  has type LIST rather than object at file root

```

Let's try this one.   Let's describe config for list of element without path to list


```scala
  val ss =
    s"""
       | [summer, winter, spring]
       |""".stripMargin

```

Same error.

If you are able to make the above two cases work using typesafe-config work (which zio-config is depending on)
 we can fix your issue. You may post the solution along with the same issues. :)

Also, feel free to fix this open issue in the above library, which will be helpful for us.
If the above issue is fixed, and if zio-config is not able to get you this feature, feel free to reimplement from the scratch :slight_smile:

**Comment by afsalthaj** [comment]
@senia-psm 
> @afsalthaj We just discussed it on discord and there is no way to implement it with current DeriveConfigDescriptor

I didn't get you. There is no way we could implement?

**Comment by senia-psm** [comment]
@afsalthaj we discussed on discord that there is no way to read `list = ["a", "c", {d: {s: "s"}}]` with `DeriveConfigDescriptor`.

**Comment by jdegoes** [comment]
@senia-psm Creating config descriptors without a path sounds like it would be useful indeed. 

Now, the tradeoff is as follows: if a config descriptor needs no path, then it's possible the user can try to read the config descriptor directly, without wrapping it into a record that has a path. In such a case, there is no way to provide any useful semantic atop property files, command-line arguments, and environment variables, as all data within them is inherently key/value based.

That said, I think it's probably a reasonable tradeoff. Next time I pair with @afsalthaj we can examine the impact on the API.

**Comment by senia-psm** [comment]
> Let's describe config for list of element without path to list, not using zio-config, but just using the basic typesafe-config library.

I don't want to use Description for list without path to parse a root node.

I'd prefer to be able to work with 2 kinds or `Description`: `Descritpion` that can parse any kind of node and `ODescription` that can parse only object nodes. Just like `Format` and `OFormat` in `play-json`.

For example `Config` in `typesafe` is just a `Map[String, ConfigValue]` and `ConfigValue` is a "kind of ADT" with 6 branches, described in `ConfigValueType` `enum`: `OBJECT`, `LIST`, `NUMBER`, `BOOLEAN`, `NULL` and `STRING`. `OBJECT` represented by `ConfigObject` which is the same as `Config`.
 
And since `zio-config` works with `String`s only (all `PropertyType` instances accept `String`) we can skip `NUMBER`, `BOOLEAN`, and`NULL`.

**Comment by senia-psm** [comment]
@jdegoes @afsalthaj please take a look at #259

**Comment by jdegoes** [comment]
Fixed by #259
